### ISSUE TITLE
Setup.py and / or pypi package

### ISSUE THREAD

**Issue by Leon0402** [ultralytics/yolov5#2525]
## 🚀 Feature
Add a setup.py or / and publish a package on pypi for easier use.

## Motivation

It is often required in ordner to load the model to have the original source code. Furthermore, there are some useful scripts in here. 
It would be a lot easier to set up a project using yolov5 if there were at least a setup.py file. Best case you would also upload it to pypi obviously.

For example: You could then use yolov5 with the package manager poetry, where you can specify a github repository. 

## Pitch

1. Add a setup.py file -> Don't have to be very advanced, should be less than 10 lines of code. Just specifying some meta data and requirements
2. Optionally upload to pypi to make it even easier to use
## Alternatives

An alternative to uploading it to pypi is just using only a setup.py. That would be a huge improvement already and it's easy to implement. 
You could obviously also use a package manager like poetry instead of a setup.py, which is easy to implement as well and can also do the uploading to pypi.

**Comment by glenn-jocher** [comment]
@Leon0402 this is a great idea! See https://github.com/ultralytics/yolov5/issues/2481 for some details on this.

**Comment by Leon0402** [comment]
Perfect! As I said, personally it would also be a a lot easier if you just provide a setup.py  -> This allows including the project in for example poetry by url. No need for pypi, although it is a nice extra :-) 

```
from setuptools import setup

setup(
    name="yolov5",
    url="https://github.com/ultralytics/yolov5",
    maintainer="ultralytics",
    maintainer_email="glenn.jocher@ultralytics.com",
    packages=["models", "utils"],
    install_requires=["opencv-python", "matplotlib", "torchvision", "PyYAML", "requests", "pandas", "seaborn"],
)
````

Might look like this. I think that's all I need to use it with poetry (at least yolov5 version4 I think?). I can make a PR, but I thought it might make more sense if you do it. 
It's mainly just listing some metatdata and you know better than me what belongs in there :-)

**Comment by glenn-jocher** [comment]
@Leon0402 hmm interesting. I hadn't thought of creating a setup.py other than for pypi. https://github.com/python-poetry looks interesting. 

Your setup.py file is very simple, are you sure that would be enough? We've been working on a similar one that's a bit more detailed here (for the TODO `ultralytics` package).
https://github.com/ultralytics/pip/blob/master/setup.py

**Comment by glenn-jocher** [comment]
@Leon0402 also could you submit a PR for your idea? I think that would be helpful to understand it better.

**Comment by Leon0402** [comment]
I'm not very experienced with setup.py or in general packaging in python. I pretty much did the minimal config I could come up with, so poetry would install all dependencies and find the packages. 

It's pretty much a subset of yours. I see just a few differences: 

1. I was just interested in those packages I actually use. As I was working on a coreml export script, I only need model/yolo.py (and that file needed a few others). 
1. You automatically read your requirements.txt, while I did list all dependencies manually without any version for only the code I include (-> Therefore my dependencies list is shorter than your requirements list)
1. I didn't care about a version
1. I didn't care about metadata, which is mainly for pypi and things like this ( I think only a few fields like maintainer were required) 

My poetry setup for yolov5 is just: 
```
[tool.poetry.dependencies]
yolov5 = { path = "../yolov5", develop = true }
```
in pyproject.toml 

So I've cloned the project and added the setup.py in the root. The develop flag is so that poetry picks up changes I do for instance switching branches on yolov5. That's also why I didn't need a version in the setup.py


If you consider using poetry: From my personal experience, poetry is a great idea and I use it in all my projects. The specification of dependencies along with automatic virtual environments just make it a lot easier for me. But it's not always super easy. It has some bugs. I also had problems with making my stuff cross plattform, especially with pytorch. It's not (completly) poetrys fault. It's mainly because different machines, different os, different python versions need different python packages and too often python packages for a specific os or python version are just missing on pypi. For instance pytorch (used to be) not available on pypi for Windows, but for all other platforms. But poetry doesn't do a good job at failing fast, giving good error messages and stuff like that. 
But still: Would recommend it!
